### PR TITLE
Validate TaskRun compatibility with the Affinity Assistant

### PR DESCRIPTION
**Pull request by jlpettersson**
# Changes

A TaskRun that mount more than one PVC-backed workspace is incompatible
with the Affinity Assistant. But there is no validation if the TaskRun
is compatible - so the TaskRun Pod is stuck with little information on why
to the user.

This commit adds validation of TaskRuns. When a TaskRun is associated with
an Affinity Assistant, it is checked that not more than one PVC workspace
is used - if so, the TaskRun will fail with a TaskRunValidationFailed condition.

Proposed in https://github.com/tektoncd/pipeline/issues/2829#issuecomment-645868118
Closes https://github.com/tektoncd/pipeline/issues/2864

/kind feature 
/release-note-none

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

### CODE REVIEW

File: pkg/reconciler/taskrun/taskrun.go [hunk 1]
```diff
@@ -700,3 +706,26 @@ func storeTaskSpec(ctx context.Context, tr *v1beta1.TaskRun, ts *v1beta1.TaskSpe
...
+
+	pvcWorkspaces := 0
+	for _, w := range tr.Spec.Workspaces {
+		if w.PersistentVolumeClaim != nil || w.VolumeClaimTemplate != nil {
+			pvcWorkspaces++
+		}
+	}
+
+	if pvcWorkspaces > 1 {
+		return fmt.Errorf("TaskRun mounts more than one PersistentVolumeClaim - that is forbidden when the Affinity Assistant is enabled")
```
Comment (sbwsg) [inline pkg/reconciler/taskrun/taskrun.go:728]: Should this be "mounts more than one writeable PVC"?  I thought it was OK to receive multiple PVCs in read-only?  Or will that also trip into potential deadlock scenarios?

Comment (jlpettersson) [reply pkg/reconciler/taskrun/taskrun.go:728]: That is a good question @sbwsg 

TLDR; as the message is written - it matches the current implementation of the Affinity Assistant.

The Affinity Assistant makes so that TaskRun pods are scheduled to the same Node. This actually helps to avoid two different problems:

1) All pods sharing a PVC (e.g. with accessMode ReadWriteOnce) can now execute in parallel - this may be possible with other accessModes e.g. ReadWriteMany even without the Affinity Assistant.
2) All pods sharing a PVC (e.g. with a zonal StorageClass) is scheduled to the same AZ, so it will not deadlock the pipeline - this may be possible with regional StorageClasses even without the Affinity Assistant - but on e.g. GCP regional storageClass volumes is only available on two of three AZs - so it may still be a problem. If using accessMode ReadWriteMany pods can execute in parallel within an AZ - but for regional clusters they additionally need to be regional StorageClass to avoid the problems that the Affinity Assistant solves.

The guidelines about "at most one writeable workspace" is for Tasks - e.g. they don't say if it is a PVC or emptyDir or regional StorageClass. With "writeable" I also meant PVCs - e.g. using a Secret or ConfigMap-workspace is still fine for Tasks - in addition to a PVC workspace.

I may add a new section to the workspace documentation to document more about this - it is a bit complicated.

Comment (jlpettersson) [reply pkg/reconciler/taskrun/taskrun.go:728]: On the other hand - using a `ReadWriteOnce` in readOnly mode and regional storageClass - should work - but is currently not supported/implemented by the Affinity Assistant. We may want to implement it - but it is a bit complicated too - and probably a rarely used use case.

File: pkg/reconciler/taskrun/taskrun.go [hunk 2]
```diff
@@ -297,6 +297,12 @@ func (c *Reconciler) prepare(ctx context.Context, tr *v1beta1.TaskRun) (*v1beta1
 		return nil, nil, controller.NewPermanentError(err)
 	}
 
+	if err := validateWorkspaceCompatibilityWithAffinityAssistant(tr); err != nil {
+		logger.Errorf("TaskRun %q workspaces are invalid: %v", tr.Name, err)
+		tr.Status.MarkResourceFailed(podconvert.ReasonFailedValidation, err)
+		return nil, nil, err
```
Comment (afrittoli) [inline pkg/reconciler/taskrun/taskrun.go:303]: This should be a `PermanentError` otherwise the controller will requeue this.

**Comment by sbwsg** [comment]
I'm going to add this to the 0.15 milestone as it would be great to get the validation in place in time for the next release.

**Comment by sbwsg** [comment]
I'm also happy to pick this up @jlpettersson if you don't have time to finish.  let me know!

File: pkg/reconciler/taskrun/taskrun.go (hunk 2 above)
Comment (jlpettersson) [reply pkg/reconciler/taskrun/taskrun.go]: Thanks! Good catch!

**Comment by jlpettersson** [comment]
> I'm also happy to pick this up @jlpettersson if you don't have time to finish. let me know!

@sbwsg I fixed the `PermanentError` - but haven't investigated if the tests can be refactored as @afrittoli suggested. Its up to you if you want to add those improvements in this PR or if it can be done in its own PR after this is merged.

Thanks @sbwsg and @afrittoli for review!

**Comment by sbwsg** [comment]
Awesome, this looks great to me as it is but I'll leave it to @afrittoli to give the PermanentError another look.

**Comment by vdemeester** [comment]
/lgtm
/hold
@afrittoli can you take a look :upside_down_face:
